whylog/config/__init__.py

```diff
@@ -0,0 +1,94 @@
+from abc import ABCMeta, abstractmethod
+
+
+class AbstractConfig(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def create_investigation_plan(self, front_input):
+        pass
+
+
+class YamlConfig(AbstractConfig):
+    def __init__(self, parsers_path, rules_path, log_locations_path):
+        pass
+
+    def create_investigation_plan(self, front_input):
+        pass
+
+    def _get_log_type(self, front_input):
+        pass
+
+    def _find_matching_parsers(self, front_input, log_type):
+        pass
+
+    def _filter_rule_set(self, parsers_list):
+        pass
+
+    def _get_locations_for_logs(self, logs_types_list):
+        pass
+
+
+class InvestigationPlan(object):
+    def __init__(self, front_input, rule_subset, log_location_dict):
+        pass
+
+    def get_next_investigation_step(self):
+        pass
+
+
+class RuleSubset(object):
+    def __init__(self, rule_dict):
+        pass
+
+    def get_logs_types(self):
+        pass
+
+    def get_rules_for_log_type(self, log_type):
+        pass
+
+    def get_parsers_for_log_type(self, log_type):
+        pass
+
+
+class InvestigationStep(object):
+    """
+    Represents rules, parsers and locations of logs which are necessary
+    to find and parse log files with potential causes.
+    """
+
+    def __init__(self, parsers, rules, log_location, effect_time):
+        pass
+
+    def get_clues(self, line):
+        """
+        Basing on parsers creates clues in investigation
+        :param line: line from parsed file
+        :returns: list of created clues
+        """
+        pass
+
+
+class Clue(object):
+    """
+    Collects all the data that parser can extract from single log line.
+    Also, contains parsed line, the location of parsed file and offset to that line.
+    """
+
+    def __init__(self, regex_parameters, line_time, line_content, file_location, file_offset):
```

Comment:
Using  file location and file offset here assumes that this properly identifies a location and context of the line. If the given file is a part of a logical file,  then this assumption is not true.

Instead, an object representing a line location, or a line source, should be passed in parameter of this method.